rbd/watchers_mimic.go

```diff
@@ -59,3 +76,77 @@ func (image *Image) ListWatchers() ([]ImageWatcher, error) {
 	}
 	return imageWatchers, nil
 }
+
+// watchCallbacks tracks the active callbacks for rbd watches
+var watchCallbacks = callbacks.New()
+
+// WatchCallback defines the function signature needed for the UpdateWatch
+// callback.
+type WatchCallback func(interface{})
+
+type watchInstance struct {
+	callback WatchCallback
+	data     interface{}
+}
+
+// Watch represents an ongoing image metadata watch.
+type Watch struct {
+	image   *Image
+	wi      watchInstance
+	handle  C.uint64_t
+	cbIndex int
+}
+
+// UpdateWatch updates the image object to watch metadata changes to the
+// image, returning a Watch object.
+//
+// Implements:
+//  int rbd_update_watch(rbd_image_t image, uint64_t *handle,
+//                       rbd_update_callback_t watch_cb, void *arg);
+func (image *Image) UpdateWatch(cb WatchCallback, data interface{}) (*Watch, error) {
+	if err := image.validate(imageIsOpen); err != nil {
+		return nil, err
+	}
+	wi := watchInstance{
+		callback: cb,
+		data:     data,
+	}
+	w := &Watch{
+		image:   image,
+		wi:      wi,
+		cbIndex: watchCallbacks.Add(wi),
+	}
+
+	ret := C.wrap_rbd_update_watch(
+		image.image,
+		&w.handle,
+		C.callWatchCallback,
+		C.uintptr_t(w.cbIndex))
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+	return w, nil
+}
+
+// Unwatch un-registers the image watch.
+//
+// Implements:
+//  int rbd_update_unwatch(rbd_image_t image, uint64_t handle);
+func (w *Watch) Unwatch() error {
+	if w.image == nil {
+		return ErrImageNotOpen
+	}
+	if err := w.image.validate(imageIsOpen); err != nil {
+		return err
```

Comment:
Sorry, but I'm unclear on what you mean here. This function is meant to be used with a defer to both clean up the C-side (rbd_update_unwatch) and the go -side (.Remove). Validate ought to only fail in the case of a programming error (calling Unwatch on an uninitialized object for example). Or if somehow the image was closed first we produce an error (but try to avoid crashing).

Reply:
Yeah, I was thinking of the second case. If the image somehow got closed, we still have the Watch in the callback registry, right?

Reply:
That entry in the registry will continue to exist, yes. Plenty of resource leaks are possible in go code if you don't make proper use of "closing" functions. os.File for example.
I don't think we ought to go out of our way to prevent anything more than really easy to trigger errors or crashes that are not recoverable by the go programmer. There should be no problems if one sticks to the normal style of call-function, check-error, defer release-function-call, like so:
```
// go pseudo code
image, err = OpenImage(...)
if err != nil {
    return err
}
defer image.Close()

w, err := image.UpdateWatch(...)
if err != nil {
    return err
}
defer w.Unwatch()

```

Reply:
Yeah, but a `watchCallbacks.Remove(w.cbIndex)` wouldn't hurt either, right? Also, does that pattern apply to typical "watching"? I would expect that these images are opened for a longer time and are not managed in a scope if you are installing an observer on them. But I'm talking out of pure assumptions. And my next thought is: wouldn't it be better in this case to link the objects the other way around, that is, linking the image to the watcher, and letting the image clean it up. Then `Unwatch()` would become an `Image` method, which one could argue fits more the C-API, since the first argument is the image pointer.